### PR TITLE
Do not trim whitespace when part of strings or regexes

### LINKED ISSUE #195010: Trim Trailing Whitespace shouldn't affect multiline strings
This is regarding the `files.trimTrailingWhitespace` setting.
While it is possible to set it to False for specific files e.g. markdown, there often cases where we have multiline strings in a code file.

for example:
```python
st.info("""
Line1  
Line2
""")
```

When `files.trimTrailingWhitespace` is set to True, it would remove the two whitespaces after Line1, which breaks their markdown rendering as two separate lines.

Would be great to have this setting ignore multilines strings.

### PR DESCRIPTION
**Pull request by 333fred**
Fixes https://github.com/microsoft/vscode/issues/195010. If the token to be removed is part of a string or a regex, then we do not want to remove the character, as it is very likely semantically important.

Before I mark this as ready to review, I need some help with how to approach testing this. Presumably, tests need to go in `trimTrailingWhitespaceCommand.test.ts`, but I'm going to need a language in the unit test suite with standard token types. Is there an example of this that isn't creating an entire language provider just for the test? Every example of `testTextModel.createTextModel` that I can see that has a language makes an entire custom one just for the test; if this is the only approach, I could use someone to point out what things I'll need to mock to test this. I would also like to add some integration tests, particularly around whitespace after the end a string, to ensure that it works with real tokenization as well.

The other open question I have is whether this behavior should be on by default, or if we need to update the trim whitespace option to have a "trim except in strings and regexes" option, as this behavior may not be desired by some?

### CODE REVIEW

**Comment by 333fred** [comment]
@hediet, since you're the assignee on #195010, would you be able to comment on my questions above?

**Comment by coreyward** [comment]
Not sure how viable this is, but I actually prefer trailing whitespace in multiline strings to be trimmed most of the time (in JS), but in some cases it's not desirable (specifically for tests). Perhaps there is a way to alter this behavior just for test files?

**Comment by 333fred** [comment]
@hediet or @alexdima, thoughts on my questions? Given the comment from @coreyward on just this draft PR, I'm definitely leaning towards "needs an option switch". Personally, I do not want whitespace trimming in a multiline string, ever, in any file, regardless of whether it's a test file or not. It could be a regex component, or part of a piece of markdown where trailing `  ` matters.

**Comment by hediet** [comment]
I think this should be behind an opt-in flag.
Also, I think when this flag is enabled, trimming whitespace should wait on the tokenization to complete.

**Comment by 333fred** [comment]
@hediet I think I've plumbed through a setting for this correctly, but I could still use some pointers on how to test this appropriately.

**Comment by 333fred** [comment]
Talked with @hediet offline. With the end of year coming up there isn't a lot of bandwidth to look at a change like this at the moment, given the potential costs of tokenization for the file. I'll ping again in January.

**Review by alexdima — CHANGES_REQUESTED** [review]

File: src/vs/editor/common/commands/trimTrailingWhitespaceCommand.ts [hunk 1]
```diff
@@ -96,6 +99,17 @@ export function trimTrailingWhitespace(model: ITextModel, cursors: Position[]):
...
 
+		if (!trimInRegexesAndStrings) {
+			model.tokenization.forceTokenization(lineNumber);
+			const lineTokens = model.tokenization.getLineTokens(lineNumber);
+			const fromColumnType = lineTokens.getStandardTokenType(lineTokens.findTokenIndexAtOffset(fromColumn));
+
+			if (fromColumnType === StandardTokenType.String || fromColumnType === StandardTokenType.RegEx) {
+				// Never trim trailing whitespace from strings and regexes, as they're likely intentional
+				continue;
+			}
```
[the remark this answers is not in the document]
Comment (hediet) [reply src/vs/editor/common/commands/trimTrailingWhitespaceCommand.ts:110]: Thinking of it, shouldn't this problem be solved by a proper (extension-powered) on-save formatter?

File: src/vs/workbench/contrib/files/browser/files.contribution.ts [hunk 2]
```diff
@@ -228,6 +228,12 @@ configurationRegistry.registerConfiguration({
 			'description': nls.localize('trimTrailingWhitespace', "When enabled, will trim trailing whitespace when saving a file."),
 			'scope': ConfigurationScope.LANGUAGE_OVERRIDABLE
 		},
+		'files.trimTrailingWhitespaceInRegexAndStrings': {
```
[the remark this answers is not in the document]
Comment (coreyward) [reply src/vs/workbench/contrib/files/browser/files.contribution.ts:231]: Yeah, they're common in Ruby.

File: src/vs/editor/common/commands/trimTrailingWhitespaceCommand.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (alexdima) [reply src/vs/editor/common/commands/trimTrailingWhitespaceCommand.ts:110]: > Thinking of it, shouldn't this problem be solved by a proper (extension-powered) on-save formatter?

Yes, but there are always a long tail of languages which don't come with language servers. So I find the feature request reasonable.

File: src/vs/workbench/contrib/files/browser/files.contribution.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (333fred) [reply src/vs/workbench/contrib/files/browser/files.contribution.ts:231]: C# regexes can also be written across multiple lines using raw strings, but I'm unsure whether this tokenizer would consider them regex or strings. Roslyn will, certainly, but it's driven with semantics and I don't know if that's surfaced here.

File: src/vs/editor/common/commands/trimTrailingWhitespaceCommand.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (333fred) [reply src/vs/editor/common/commands/trimTrailingWhitespaceCommand.ts:110]: > If the model has tokens, then we consult them, if it doesn't have tokens, we don't trim the trailing whitespace.

This approach seems fine to me, but do we need to clarify the setting description to convey this info?

[the remark this answers is not in the document]
Comment (333fred) [reply src/vs/editor/common/commands/trimTrailingWhitespaceCommand.ts:115]: Updated. Still looking for advice on how to actually test all of this though.

**Comment by hediet** [comment]
For testing, please see existing unit tests for tokenization. Basically you should create a fake language with a static tokenizer where you set tokens manually. Then look for tests that test TrimTrailingWhitespaceCommand (or other commands) and try to run the command on the text model with the fake language.

**Review by alexdima — APPROVED** [review]
Thank you!

**Review by Tyriar — APPROVED** [review]

**Comment by 333fred** [comment]
@alexdima thanks for picking this up! Really appreciate it, I've been very busy with other things and was unlikely to get back to this for another few months.
